colossalai/zero/zero_optimizer.py

```diff
@@ -197,24 +183,31 @@ def _maybe_move_fp32_params(self):
             available_cuda_margin_mem = self.gemini_manager.cuda_margin_mem * self.gpu_margin_mem_ratio
             fp32_params_available_cuda_margin_mem = available_cuda_margin_mem / self.optim.num_fp32_shards_per_param
             fp32_params_used_cuda_margin_mem = 0
-            for fp16_param_chunk, fp32_param_chunk in zip(self.chunk_manager.chunk_groups['fp16_param'],
-                                                          self.chunk_manager.chunk_groups['fp32_param']):
-                if fp32_param_chunk.is_empty:
-                    continue
-                if fp32_params_used_cuda_margin_mem + fp32_param_chunk.mem < fp32_params_available_cuda_margin_mem:
-                    self.chunk_manager.move_chunk(fp32_param_chunk, get_current_device())
-                    # stores grad now
-                    self.chunk_manager.move_chunk(fp16_param_chunk, get_current_device())
-                    self.module._set_chunk_grad_device(fp16_param_chunk, get_current_device())
-                    fp32_params_used_cuda_margin_mem += fp32_param_chunk.mem
-                    for p in fp16_param_chunk.get_tensors():
-                        state = self.optim.state[p]
+
+            for group in self.param_groups:
+                for fake_param in group['params']:
+                    chunk32 = self.param_to_chunk32[fake_param]
```

Comment:
Shall we further filter those parameters with no grad? As we assume that it's a static graph, those parameters are not required to optimized.

Reply:
Do you mean unused parameters? In my opinion, a parameter should have its gradient.

Reply:
When finetuning, some parameters may be frozen. They are used during forward pass, but they have no grad.

Reply:
currently, it doesn't support this feature. but users can set the learning rate of these parameters to 0 instead.